File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] ans = {-1, -1};
        int left = LEFT(nums, target);
        int right = RIGHT(nums, target);
        ans[0] = left;
        ans[1] = right;
        return ans;
    }
    public int LEFT(int[] arr, int target) {
        int low = 0;
        int high = arr.length-1;
        int first = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            // == condition
            if (arr[mid] == target) {
                first = mid;
                high = mid - 1;
            }
            // < condition
            else if (arr[mid] < target) {
                low = mid + 1;
            }
            // > condition
            else {
                high = mid - 1;
            }
        }
        return first;
    }

    public int RIGHT(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1;
        int last = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            // == condition
            if (arr[mid] == target) {
                last = mid;
                low = mid + 1;
            }
            // < condition
            else if (arr[mid] < target) {
                low = mid + 1;
            }
            // > condition
            else {
                high = mid - 1;
            }
        }
        return last;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int findLeftMost(vector<int>& nums, int target, int n) {
        int s = 0;
        int e = n-1;
        int left_most = -1;
        while (s <= e) {
            int mid = s + (e-s)/2;
            if (target == nums[mid]) {
                left_most = mid;
                e = mid - 1;
            } else if (target > nums[mid]) {
                s = mid + 1;
            } else {
                e = mid - 1;
            }
        }
        return left_most;
    }
    int findRightMost(vector<int>& nums, int target, int n) {
        int s = 0;
        int e = n-1;
        int right_most = -1;
        while (s <= e) {
            int mid = s + (e-s)/2;
            if (target == nums[mid]) {
                right_most = mid;
                s = mid + 1;
            } else if (target > nums[mid]) {
                s = mid + 1;
            } else {
                e = mid - 1;
            }
        }
        return right_most;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int left_most = findLeftMost(nums, target, n);
        int right_most = findRightMost(nums, target, n);
        return {left_most, right_most};
    }
};